packages/flutter/lib/src/widgets/navigator.dart

```diff
@@ -3941,5 +4017,15 @@ class NavigatorState extends State<Navigator> with TickerProviderStateMixin {
         ),
       ),
     );
+
+    if (_heroControllerFromScope != null) {
+      // Hides the HeroControllerScope for the widget subtree so that the other
+      // nested navigator underneath will not pick up the hero controller above
+      // this level.
+      subtree = HeroControllerScope(
+        child: subtree
+      );
+    }
```

Comment:
Why not always wrap it?

Reply:
This is an optimization I feel like doing. Would you rather be safe and always wrap it?

Reply:
Let's always wrap it to avoid changing the shape of the tree when this changes.

Reply:
Ah yes, I forgot about changing the tree shape cause the whole subtree to rebuild...